### DISCUSSION TITLE
External dataset  Channel

### DISCUSSION THREAD

**Discussion by MarHajj** [NVlabs/sionna#47] [Q&A] [answered]
Hi,

I have an external dataset channel done by measurements and simulations (amplitude, phase and delays of paths) and i need to introduce this data into sionna in aim to simulate the BER.

I don't understand what number of time steps means? and why the number of time steps should be equal to the number of OFDM symbols in the Resource Grid.

Thank you

**Comment by jhoydis** [comment, marked as answer]
We assume that each example consists of a sequence of time varying channel impulse responses. 

For example, in the 3GPP 38.901 channel models, users are dropped within a cell and then move along a trajectory with a given speed.
The channel impulse responses are then sampled along this trajectory for ``num_time_steps``.

If your dataset does not have any form of time evolution, the number of time steps would be simply one.

However, if you want to use use OFDM with a resource grid you need to provide one channel impulse response realization for each OFDM symbol in the resource grid. This means that you would need to set ``num_time_steps`` equal to the number of OFDM symbols in the resource grid and assign the same value of the channel impulse response  to each of the time steps.

**Comment by victorcionca** [comment]
Hi all,

I also have a question on the topic of external datasets. I am using the NIST Industrial Wireless dataset https://www.nist.gov/ctl/smart-connected-systems-division/networked-control-systems-group/measurement-data-files. This is provided as CIRs, from which I am extracting the multi-path components and their delays using peak detection, and normalizing the result so that total power equals 1.

For the number of time samples I have used the resource_grid.num_time_samples + l_total - 1 as per the API documentation. However, the resource_grid.num_time_samples is not equal to the number of OFDM symbols, it is equal to number of OFDM symbols * (FFT bins + CP). Considering the shape of the CIR tensor as (batchsize, num_rx, num_rx_ant, num_tx, num_tx_ant, num_paths, num_timesteps+l_tot-1), the time steps represent the evolution in time of a single MPC over all the timesteps of the RG + l_total. Is my understanding correct?

Nevertheless I was unsure of how to map the dataset CIRs to the number of timesteps and I ended up tiling the path coefficients for all timesteps, which implies that the channel response is constant throughout the RG.

Not sure if that is correct, but the results I am getting are very pessimistic. I am using an OFDM PHY with 5G polar code and, for a channel with LoS and k factor 10 I am getting BER approx 1e-2 at 15dB, whereas on a flat fading Rice channel model with same k factor I am getting closer to 1e-4 at 15dB. I have tried several other channels but the results are similar. The BER does decrease with SNR reaching 1e-5 at 35dB. I don't know whether the channel is really that poor (which shouldn't based on the k factor) or my modelling is incorrect.

Hope you can shed some light.

Best regards,
Victor

**Reply by jhoydis** [reply]
Hi Victor,

If you want to make simulations in the frequency domain using the [OFDMChannel](https://nvlabs.github.io/sionna/api/channel.html#sionna.channel.OFDMChannel), your CIR you needs as many time steps as OFDM symbols in a resource grid.  This means that each OFDM symbols gets a different CIR.

If you want to do your simulations in the time domain using the [TimeChannel](https://nvlabs.github.io/sionna/api/channel.html#sionna.channel.TimeChannel), your CIR needs to have ``resource_grid.num_time_samples + l_total - 1`` time steps. This means that the CIR might change within the duration of an OFDM symbol.

Unless you want to capture a scenario with really high Doppler spreads or insufficient cyclic prefix length, you should make frequency domain simulations.

If your channel is constant across the entire resource grid, channel estimation should be fairly easy and I cannot explain from the information you have provided why the performance is so poor. Do you simulate a MIMO setup or is it SISO? Do you estimate the channels or assume perfect CSI? Did you activate the ``normalize_channel=True`` flag of the Time/OFDMChannel?

**Reply by victorcionca** [reply]
Hi Jakob,

Many thanks for the quick answer and advice. And even more thanks to yourself and the entire team for developing and releasing Sionna. It is really an amazing tool!

I am using SISO and because I am normalizing manually the CIRs I did not set the normalize_channel flag. However, I will try that instead, see how it works.

I am using channel estimation (Least Squares) and equalization (LMMSE) as per one of the tutorials.

Initially we were looking at variable CP length and its impact on ISI, hence the use of the Time Channel. However, very good advice to switch to the frequency domain. I will try that as well.

Many thanks!
Victor

**Comment by MarHajj** [comment]
Hi again,
For a defined scenario (single transmitter, single receiver and SISO antennas), I have 112 realizations of channel (each one contains amplitude (dB) and phase(radian) and delays(ns) of each paths) done by simulation. I will use a time steps of 14. 
What is the number of data set size in my case? Is-it 112 (equal to the number of realizations) ?
Thank you,
M. EL HAJJ

**Reply by victorcionca** [reply]
Hi @MarHajj 

If you're using an OFDM channel you need one realization per OFDM symbol used in your OFDM grid.
You have to implement a CIR generator as per the documentation, which will return the amplitude of the paths and their delays as two tensors.
You have to decide how many sample points you will use from each CIR. The documentation talks about paths (MPCs). I tried to use just the peaks from the CIR, but the results were very poor. Eventually I just used all the amplitude samples from the CIR, in my case max 60. 

Considering your scenario the shape of the amplitude tensor should be (1,1,1,1, num_paths (or num samples), num_ofdm_symbols).
Then you have the delay tensor whose shape should be (1,1,num_paths (or num samples)).

I'm very curious to see your results. See below for mine for an LDPC coded PHY, two separate datasets, one with an average Rice K factor of ~10 the other ~5.
![image](https://user-images.githubusercontent.com/12625899/192104848-700ff5fe-6604-47d9-8bac-41b16609741b.png)
(Edited as found an issue in the simulation)

Best regards,
Victor

**Reply by MarHajj** [reply]
hi @victorcionca 
Yes, i am using an OFDM channel (14 OFDM symbol in the resource grid with numb of paths = 45). The shape of the amplitude tensor miss the dataset_size. 
Amplitude tensor shape [dataset_size, num_rx, num_rx_ant, num_tx, num_tx_ant, num_paths, num_time_steps] and my question concern the dataset_size. 

Best regards,
Marwan

**Reply by victorcionca** [reply]
Hi Marwan,
It depends on how you load the CIRs into the channel model. If you use ``cir_to_ofdm_channel`` the first parameter is not the dataset size but the batch size. So what you should probably do there is sample from your dataset ``batch_size`` CIRs then convert them into a tensor with the shape ``batch_size, 1, 1, 1, 1, 45, 14``.

If you use ``CIRDataset`` you need to implement a generator class that will return single instances. ``CIRDataset`` will then take those instances, shuffle and tile them to cover all MC iterations.

I think that if your dataset represents stationary measurements you can use either. If you have mobile measurements you should use ``cir_to_ofdm_channel`` so that you can sample accordingly, making sure that the RG is covered by a consecutive set of CIRs.

I hope this makes sense.

Best,
Victor

**Reply by yhuwhai** [reply]
Hi @MarHajj , 

I followed your discussion.

I am wondering if you were successful with training the Neural SIMO receiver with the external dataset.

I have tried using a ray tracing generated CIRs though the results are not so optimistic.

Thanks and best regards,
Uyoata
